src/components/TimePicker/TimePicker.js

```diff
@@ -107,73 +107,90 @@ function TimePicker({forwardedRef, defaultValue, onSubmit, onInputChange}) {
+                */
+                newHour = `${firstDigit}${firstDigit === '1' && secondDigit > 1 ? 0 : secondDigit}`;
                 newSelection = 1;
-            }
-        } else if (selectionHour.start === 1) {
-            // Case when the cursor is at the second position.
-            const formattedText = `${filteredText[0]}${filteredText[1]}`;
-
-            if (filteredText.length < 2) {
-                // If we remove a value, prepend 0.
-                newHour = `0${text}`;
-                newSelection = 0;
-                // If the second digit is > 2, replace the hour with 0 and the second digit.
-            } else if (formattedText > 12 && formattedText <= 24) {
-                newHour = String(formattedText - 12).padStart(2, '0');
-                newSelection = 2;
-                setAmPmValue(CONST.TIME_PERIOD.PM);
-            } else if (formattedText > 24) {
-                newHour = `0${text[1]}`;
-                newSelection = 2;
             } else {
-                newHour = `${text[0]}${text[1]}`;
-                setHours(newHour);
+                // The first entered digit is 2-9. We should replace the whole value by prepending 0 to the entered digit.
+                newHour = `0${firstDigit}`;
                 newSelection = 2;
             }
-        } else if (selectionHour.start === 2 && selectionHour.end === 2) {
-            // Case when the cursor is at the end and no text is selected.
-            if (filteredText.length < 2) {
-                newHour = `${text}0`;
-                newSelection = 1;
+        } else if (selectionHour.start === 1 && selectionHour.end === 1) {
+            // The cursor is in-between the digits
+            if (trimmedText.length < 2) {
+                // We have removed the first digit. Replace it with 0 and move the cursor to the start.
+                newHour = `0${trimmedText}`;
+                newSelection = 0;
             } else {
+                newHour = `${trimmedText[0]}${trimmedText[1] || 0}`;
                 newSelection = 2;
             }
+        } else if (trimmedText.length <= 1 && trimmedText <= 1) {
+            /*
+             The trimmed text is either 0 or 1.
+             We are either replacing hours with a single digit, or removing the last digit.
+             In both cases, we should append 0 to the remaining value.
+             Note: we must check the length of the filtered text to avoid incorrectly handling e.g. "01" as "1".
+            */
+            newHour = `${trimmedText}0`;
+            newSelection = 1;
+        } else {
+            newHour = trimmedText;
+            newSelection = 2;
+        }
+
+        if (newHour > 24) {
+            newHour = hours;
+        } else if (newHour > 12) {
+            newHour = String(newHour - 12).padStart(2, '0');
+            setAmPmValue(CONST.TIME_PERIOD.PM);
+        } else {
+            newHour = newHour.padStart(2, '0');
```

Comment:
```suggestion
        } else {
            newHour = trimmedText.padStart(2, '0');;
            newSelection = 2;
        }

        if (newHour > 24) {
            newHour = hours;
        } else if (newHour > 12) {
            newHour = String(newHour - 12).padStart(2, '0');
            setAmPmValue(CONST.TIME_PERIOD.PM);
```

I think the padStart is only required for that specific case. Let's do the same with the minutes